### PR TITLE
raftstore: check log entries for prepare merge based on minimal committed

### PR DESCRIPTION
**Pull request by NingLin-P**
Signed-off-by: linning <linningde25@gmail.com>

### What is changed and how it works?

What's Changed:

Check log entries for prepare merge based on minimal committed instead of minimal matched

Tests 

- Unit test

### Release note 
- raftstore: check log entries for prepare merge based on minimal committed

### CODE REVIEW

File: components/raftstore/src/store/peer.rs [hunk 1]
```diff
@@ -2233,18 +2235,25 @@ impl Peer {
...
             ));
         }
         let mut entry_size = 0;
-        for entry in self.raft_group.raft.raft_log.entries(min_index, NO_LIMIT)? {
-            entry_size += entry.get_data().len();
+        for entry in self
+            .raft_group
+            .raft
+            .raft_log
+            .entries(min_committed + 1, NO_LIMIT)?
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:2251]: Any tests?

File: Cargo.toml [hunk 2]
```diff
@@ -175,8 +175,8 @@ zipf = "6.1.0"
 
 [patch.crates-io]
 # TODO: remove this when new raft-rs is published.
-raft = { git = "https://github.com/pingcap/raft-rs", branch = "master", default-features = false }
-raft-proto = { git = "https://github.com/pingcap/raft-rs", branch = "master", default-features = false }
+raft = { git = "https://github.com/NingLin-P/raft-rs", branch = "commit-progress", default-features = false }
+raft-proto = { git = "https://github.com/NingLin-P/raft-rs", branch = "commit-progress", default-features = false }
```
Comment (NingLin-P) [inline Cargo.toml:179]: Wait https://github.com/tikv/raft-rs/pull/375 to merge.

File: components/raftstore/src/store/peer.rs [hunk 3]
```diff
@@ -2233,18 +2235,28 @@ impl Peer {
         req: &mut RaftCmdRequest,
     ) -> Result<()> {
         let last_index = self.raft_group.raft.raft_log.last_index();
-        let min_progress = self.get_min_progress()?;
-        let min_index = min_progress + 1;
-        if min_progress == 0 || last_index - min_progress > ctx.cfg.merge_max_log_gap {
+        let (min_matched, min_committed) = self.get_min_progress()?;
+        if min_matched == 0
+            || min_committed == 0
+            || last_index - min_matched > ctx.cfg.merge_max_log_gap
```
Comment (gengliqi) [inline components/raftstore/src/store/peer.rs:2241]: I think the `last_index - min_committed` also needs to be controlled. For example, how about `2*merge_max_log_gap`?

File: components/raftstore/src/store/peer.rs (hunk 1 above)
Comment (NingLin-P) [reply components/raftstore/src/store/peer.rs:2254]: Added, PTAL.

**Comment by Little-Wallace** [comment]
Is there any issue that this PR will fix?

**Comment by gengliqi** [comment]
> Is there any issue that this PR will fix?

This PR fixes a merge recover bug when network partition and maybe reduces some bugs in corner cases which I didn't find yet.
The test which reproduces this bug can be written but it is incompatible to this PR because this PR enhances the condition of proposing `PrepareMerge` which makes the merge process failed in this test.
So I think to add a test that just checks this new condition is enough.

File: Cargo.toml (hunk 2 above)
Comment (NingLin-P) [reply Cargo.toml]: Done.

File: components/raftstore/src/store/peer.rs [hunk 4]
```diff
@@ -2233,18 +2235,29 @@ impl Peer {
         req: &mut RaftCmdRequest,
     ) -> Result<()> {
         let last_index = self.raft_group.raft.raft_log.last_index();
-        let min_progress = self.get_min_progress()?;
-        let min_index = min_progress + 1;
-        if min_progress == 0 || last_index - min_progress > ctx.cfg.merge_max_log_gap {
+        let (min_matched, min_committed) = self.get_min_progress()?;
```
Comment (gengliqi) [inline components/raftstore/src/store/peer.rs:2238]: The correctness depends on `min_matched` must be greater than or equal to `min_committed`.
Although it is right, I think it's better to use assert or panic to defend some unexpected fault.

File: components/raftstore/src/store/peer.rs [hunk 5]
```diff
@@ -2199,9 +2199,11 @@ impl Peer {
         true
     }
 
-    // For now, it is only used in merge.
-    pub fn get_min_progress(&self) -> Result<u64> {
-        let mut min = None;
+    /// Return (minimal matched, minimal committed_index)
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:2202]: s/Return/Returns/

File: components/raftstore/src/store/peer.rs [hunk 6]
```diff
@@ -2233,18 +2235,30 @@ impl Peer {
         req: &mut RaftCmdRequest,
     ) -> Result<()> {
         let last_index = self.raft_group.raft.raft_log.last_index();
-        let min_progress = self.get_min_progress()?;
-        let min_index = min_progress + 1;
-        if min_progress == 0 || last_index - min_progress > ctx.cfg.merge_max_log_gap {
+        let (min_matched, min_committed) = self.get_min_progress()?;
+        assert!(min_matched >= min_committed);
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:2239]: `min_matched` can be reset to 0, however `min_committed` won't be. The assert should be moved to L2251.

File: components/raftstore/src/store/peer.rs [hunk 7]
```diff
@@ -2233,18 +2235,30 @@ impl Peer {
...
+        if min_matched == 0
+            || min_committed == 0
+            || last_index - min_matched > ctx.cfg.merge_max_log_gap
+            || last_index - min_committed > ctx.cfg.merge_max_log_gap * 2
+        {
             return Err(box_err!(
                 "log gap ({}, {}] is too large, skip merge",
-                min_progress,
+                min_matched,
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:2247]: Should include `min_committed`.

File: tests/integrations/raftstore/test_merge.rs [hunk 8]
```diff
@@ -213,10 +213,20 @@ fn test_node_merge_prerequisites_check() {
...
     let right_on_store1 = find_peer(&right, 1).unwrap().to_owned();
     cluster.must_transfer_leader(right.get_id(), right_on_store1);
-    cluster.add_send_filter(CloneFilterFactory(RegionPacketFilter::new(
-        left.get_id(),
-        3,
-    )));
+
+    // first MsgAppend will append log, second MsgAppend will set commit index,
+    // So only allowing first MsgAppend to make source peer have uncommitted entries.
+    let append_filter = RegionPacketFilter::new(left.get_id(), 3)
```
Comment (BusyJay) [inline tests/integrations/raftstore/test_merge.rs:219]: Now the case only checks committed index.

Comment (NingLin-P) [reply tests/integrations/raftstore/test_merge.rs]: Fixed, PTAL

File: tests/integrations/raftstore/test_merge.rs [hunk 9]
```diff
@@ -213,20 +213,31 @@ fn test_node_merge_prerequisites_check() {
...
-    // log gap contains admin entries.
+    // log gap (min_committed, last_index] contains admin entries.
     assert!(res.get_header().has_error(), "{:?}", res);
     cluster.clear_send_filters();
     cluster.must_put(b"k22", b"v22");
     must_get_equal(&cluster.get_engine(3), b"k22", b"v22");
 
     cluster.add_send_filter(CloneFilterFactory(RegionPacketFilter::new(
-        right.get_id(),
+        left.get_id(),
```
Comment (BusyJay) [inline tests/integrations/raftstore/test_merge.rs:240]: Why `left`?

Comment (NingLin-P) [reply tests/integrations/raftstore/test_merge.rs]: Sorry, it was a mistake.

**Pull request merged**